radicle-node/src/service.rs

```diff
@@ -1433,6 +1436,33 @@ where
             self.connect(id, addr.clone());
         }
     }
+
+    /// Maintain persistent peer connections.
+    fn maintain_persistent(&mut self) {
+        let now = self.local_time();
+        let mut reconnect = Vec::new();
+
+        for (nid, session) in self.sessions.disconnected_mut() {
+            if let Some(addr) = self.config.peer(nid) {
+                if let session::State::Disconnected { retry_at, .. } = &mut session.state {
+                    // TODO: Try to reconnect only if the peer was attempted. A disconnect without
+                    // even a successful attempt means that we're unlikely to be able to reconnect.
+
+                    if now >= *retry_at {
+                        // FIXME: Make sure we don't attempt two concurrent outgoing connections
+                        // to the same peer.
+                        reconnect.push((*nid, addr.clone(), session.attempts()));
```

Comment:
Could you easily ensure this by using:
```rust
let mut reconnect = BTreeMap<NodeId, (Addr, usize)>::new();
// ...
reconnect.insert(*nid, (addr.clone(), session.attempts()));
```

Reply:
That's not the problem actually, the  problem is if this function gets called multiple times in a row; or right after a `Command::Connect` is received.

Reply:
Hence the state of what is being attempted needs to be service-global and can't be local to this function.